packages/lodestar/src/chain/stateCache/stateContextCache.ts

```diff
@@ -1,16 +1,27 @@
 import {ByteVector, toHexString, TreeBacked} from "@chainsafe/ssz";
-import {BeaconState} from "@chainsafe/lodestar-types";
+import {BeaconState, Epoch} from "@chainsafe/lodestar-types";
 import {ITreeStateContext} from "../interface";
 
+const MAX_STATES = 96;
+
 /**
  * In memory cache of BeaconState and connected EpochContext
  *
  * Similar API to Repository
  */
 export class StateContextCache {
   private cache: Record<string, ITreeStateContext>;
-  constructor() {
+  /**
+   * Epoch -> Set<blockRoot>
+   */
+  private epochIndex: Record<Epoch, Set<string>>;
+
+  private maxState: number;
+
+  constructor(maxState = MAX_STATES) {
     this.cache = {};
+    this.epochIndex = {};
```

Comment:
Should we use a plain object or a map?

Reply:
I don't have a preference for it, just do consistently with other caches but this is a good point.

  For `cache`, maybe using `Map` is better than `Record` since looping through keys respect the insertion order in `Map` while `Object.keys(Record)` is not

```
const cache: Record<string, string> = {};
    cache["1"] = "1";
    cache["3"] = "3";
    cache["5"] = "5";
    cache["2"] = "2";
    cache["4"] = "4";
    // this proves Record does not respect insertion order
    expect(Object.keys(cache)).to.be.deep.equal(["1", "2", "3", "4", "5"]);
```

could you confirm @wemeetagain

Reply:
Map and Record both respect insertion order. From that perspective, it doesn't really matter.
I think the only place it matters is the type interface. I think Map has a 'safer' interface in that it can return undefined. IMO it doesn't really matter here as long as the implementation is correct tho :joy:

Reply:
but the test above shows that `Object.keys(Record)` does not respect insertion order

Reply:
the order inserted is 1 3 5 2 4, not 1 2 3 4 5

Reply:
the order inserted is 1 3 5 2 4 and Object.keys(Record) shows 1 2 3 4 5, we want it to be 1 3 5 2 4 too

Reply:
Oh I see... learn something new every day :thinking: 
I'm not too sure it matters for this purpose? Are we using the order somewhere?
I thought either way we'd have some code that looks like `deleteAllBeforeEpoch`

Reply:
it matters in `prune` flow https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar/src/chain/stateCache/stateContextCache.ts#L56

I think let's switch to using `Map` instead

Reply:
looks like object uses the following rules:
    First, the keys that are integer indices (what these are is explained later), in ascending numeric order.
    Then, all other string keys, in the order in which they were added to the object.
    Lastly, all symbol keys, in the order in which they were added to the object.

Reply:
According to those rules, we likely don't need to change the cache implementation here. If we want to refactor both of the state caches in a future PR, go for it.

Reply:
agree